truffle/src/main/java/org/jruby/truffle/nodes/dispatch/DispatchNode.java

```diff
@@ -136,5 +158,82 @@ public final Object execute(VirtualFrame frame) {
     public DispatchAction getDispatchAction() {
         return dispatchAction;
     }
+    
+    private Object[] splat(Object argument) {
+        // TODO(CS): what happens if isn't just one argument, or it isn't an Array?
+
+        if (!(argument instanceof RubyArray)) {
+            splatNotArrayProfile.enter();
+            notDesignedForCompilation();
+            throw new UnsupportedOperationException();
+        }
+
+        final RubyArray array = (RubyArray) argument;
+        final int size = array.getSize();
+        final Object store = array.getStore();
+
+        if (seenNullInUnsplat && store == null) {
+            return new Object[]{};
+        } else if (seenIntegerFixnumInUnsplat && store instanceof int[]) {
+            return ArrayUtils.boxUntil((int[]) store, size);
+        } else if (seenLongFixnumInUnsplat && store instanceof long[]) {
+            return ArrayUtils.boxUntil((long[]) store, size);
+        } else if (seenObjectInUnsplat && store instanceof Object[]) {
+            return ArrayUtils.extractRange((Object[]) store, 0, size);
+        }
+
+        CompilerDirectives.transferToInterpreterAndInvalidate();
+
+        if (store == null) {
+            seenNullInUnsplat = true;
+            return new Object[]{};
+        } else if (store instanceof int[]) {
+            seenIntegerFixnumInUnsplat = true;
+            return ArrayUtils.boxUntil((int[]) store, size);
+        } else if (store instanceof long[]) {
+            seenLongFixnumInUnsplat = true;
+            return ArrayUtils.boxUntil((long[]) store, size);
+        } else if (store instanceof Object[]) {
+            seenObjectInUnsplat = true;
+            return ArrayUtils.extractRange((Object[]) store, 0, size);
+        }
+
+        throw new UnsupportedOperationException();
+    }
+    
+    @ExplodeLoop
+    protected Object executeArguments(VirtualFrame frame, Object argumentOverride) {
```

Comment:
I am wondering whether there's a better way to implement this. Argument nodes for calls are now stored in `DispatchNode`, but there are still places where argument objects (no nodes, but objects) are passed when calling the dispatch head node (e.g. `ArrayNodes.inject`). That's why we are keeping this option (not storing the argument nodes in `DispatchNode` but passing objects as before) still open.